### PR TITLE
Registration Feature

### PR DESCRIPTION
**Pull request by vincentluague**
Using devise for registration
Add Registration UI
NavBar Ui
image

Rspec testing for model
image

Mailer confirmation via Gmail
image

Validation message when successfully registered
image

### CODE REVIEW

File: app/assets/stylesheets/application.css [hunk 1]
```diff
@@ -13,3 +13,51 @@
...
+}
+
+.form-control{
+  font-size: 1.4rem !important;
+}
+.navbar-brand{
+  color: white !important;
+  font-family: 'Rubik', sans-serif !important;
+  font-weight: 800;
+}
```
Comment (analyn-cajocson) [inline app/assets/stylesheets/application.css]: add new line at the end of file

File: app/views/components/_navbar.html.erb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  <div class="collapse navbar-collapse" id="navbarNavAltMarkup">
+    <div class="navbar-nav ml-auto ">
+
+        <%= link_to root_path, class:"nav-link nav-item text-white" do %>Home <span class="sr-only">(current)</span><%end%>
+        <%= link_to new_user_session_path, class:"nav-link nav-item text-white" do %>Sign In <span class="sr-only">(current)</span><%end%>
+        <%= link_to new_user_registration_path, class:"nav-link nav-item text-white" do %>Sign Up <span class="sr-only">(current)</span><%end%>
+      
+    </div>
+  </div>
+</nav>
```
Comment (analyn-cajocson) [inline app/views/components/_navbar.html.erb]: add new line

File: app/views/components/_validations.html.erb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+					Error
+				</div>
+					<div class="alert-description text-sm text-red-600">
+						<%= alert %>
+				</div>
+			</div>
+		</div>
+	</div>
+<%end%>
+
```
Comment (analyn-cajocson) [inline app/views/components/_validations.html.erb]: Fix code alignment: 2spaces

File: app/views/user_mailer/welcome_email.hmtl.erb [hunk 4]
```diff
@@ -0,0 +1,3 @@
+<h1>Welcome to the Stock Trading App, <%= @user.email %></h1>
+<p>You have successfully approved your signup. Thankyou for joining Stock Trading App <%=@user.username%></p>
+<p>You can now signin your credentials. Enjoy and have a great day :) </p>
```
Comment (analyn-cajocson) [inline app/views/user_mailer/welcome_email.hmtl.erb]: add new line

File: app/views/user_mailer/welcome_email.text.erb [hunk 5]
```diff
@@ -0,0 +1,8 @@
+Welcome to the Stock Transfer App, <%#= @user.first_name %>
+===============================================
+
+You have successfully approved your signup. Thankyou for joining Stock Trading App <%=@user.username%>
+
+Thanks for joining and have a great day!
+
+You can now signin your credentials. Enjoy and have a great day :)
```
Comment (analyn-cajocson) [inline app/views/user_mailer/welcome_email.text.erb]: add new line

File: config/environments/development.rb [hunk 6]
```diff
@@ -59,4 +59,16 @@
...
+  
+  config.action_mailer.default_url_options = { host: 'localhost', port: 3000 }
+  config.action_mailer.perform_deliveries = true
+  config.action_mailer.delivery_method = :smtp
+  config.action_mailer.smtp_settings = {
+  address:              'smtp.gmail.com',
+  port:                 587,
+  domain:               'gmail.com',
+  user_name:            'jrld.sample@gmail.com',
+  password:             'gtzrsqqodhvepnmx',
```
Comment (analyn-cajocson) [inline config/environments/development.rb]: Do not push your credentials here

File: config/routes.rb [hunk 7]
```diff
@@ -1,3 +1,11 @@
...
-  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  devise_for :users, path: '/', path_names: { sign_in: 'signin',  sign_up: 'signup', sign_out: 'signout' }, controllers: { registrations: "users/registrations", confirmations: 'confirmations', sessions: 'users/sessions' }
+  devise_scope :user do
+    post '/signin',         to: 'users/sessions#create',       as: 'signin_create'
+    post '/signup',         to: 'users/registrations#create',  as: 'signup_create'
+    delete '/signout',      to: 'devise/sessions#destroy',      as: 'signout'
+  end
+  
+  root to: "home#index"
+get '/home', to: "home#home"
```
Comment (analyn-cajocson) [inline config/routes.rb]: Fix code alignment

File: spec/models/user_spec.rb [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+    it { expect(user2).not_to be_valid }
+  end
+
+  describe 'length validations' do
+    it 'is not allow a firstname longer than 20 character' do
+      user.firstname = 'test' * 200
+      expect(user).not_to be_valid
+    end
+  end
```
Comment (analyn-cajocson) [inline spec/models/user_spec.rb]: I think you can include these spec tests to `context 'with validation'`

File: spec/models/user_spec.rb [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+  it 'is role of trader' do
+    user.role = 1
+    expect(user).not_to be_valid
+  end
+
+  it 'is role of admin' do
+    user.role = 0
+    expect(user).not_to be_valid
+  end
```
Comment (analyn-cajocson) [inline spec/models/user_spec.rb]: you can set up another `user` for these specs

**Review by analyn-cajocson — CHANGES_REQUESTED** [review]

File: app/views/components/_validations.html.erb (hunk 3 above)
Comment (vincentluague) [reply app/views/components/_validations.html.erb]: Mam naka 2 spaces naman po sya ewan ko lang po bat ganyan sa code po.
image

**Pull request merged**